### PR TITLE
[model, data, ci] refactor: collapse multimodal metadata to a single grid_thw key

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

Follow-up to #772 (multimodal forward metadata precompute). Addresses two
review TODOs from that PR: the redundant dual representation of the grid
metadata, and the per-modality `vit_*` kwarg threading.

**Item 1 — single key, no per-sample list sidecar.**
Previously the data transform called `per_sample_metadata()` to `.tolist()`
each sample's `image_grid_thw` / `video_grid_thw` CPU tensor into a parallel
`*_grid_thw_list` Python list, and `PackingCollator` special-cased flattening
those lists across the batch — a second representation of information already
carried by the `*_grid_thw` tensors. Now:

- `veomni/data/mm_metadata.py` is deleted. `per_sample_metadata` is gone;
  `merge_position_id_returns` is inlined into `data_transform.py` as a plain
  `position_ids` copy.
- The `_process_sample_qwen_vl_base` / `process_sample_qwen_omni` transforms
  emit only the HF processor's `image_grid_thw` / `video_grid_thw` CPU
  tensors — which already flow through the collator via the existing
  `DataCollateInfo` (`pack_dim=0` `torch.cat`).
- `PackingCollator` drops the `*_grid_thw_list` special-case flatten.
- Each model's `collate_multimodal_metadata` hook reads the packed
  `batch["image_grid_thw"]` tensor and does `.tolist()` once (CPU, inside the
  dataloader worker — no host-device sync). The grid tensor stays in the
  batch as the HF `get_image_features` arg; the hook only reads it.

**Item 2 — one `vit_metadata` sub-dict instead of 3 kwargs.**
Previously `Model.forward` unpacked `multimodal_metadata` into 3 separate
per-modality kwargs (`vit_grid_thw_list` / `vit_cu_seqlens` / `vit_max_seqlen`)
threaded to `get_image_features` / `get_video_features` → `ViT.forward`. Now
`Model.forward` bundles them into one `vit_metadata` sub-dict per modality;
`ViT.forward` pops the single `vit_metadata` kwarg. `dummy_forward` (FSDP
path) builds the same sub-dict. `qwen2_5_vl` / `qwen2_5_omni` (fallback-only,
window-attention ViT) drain it with one `pop`.

The `multimodal_metadata` top-level dict contract is unchanged.

### Checklist Before Starting

- Follow-up to #772 ("Follow-up TODO" items 1 + 2).
- PR title follows `[{modules}] {type}: {description}` format.

### Test

Validated on an A100 worker:

- `pytest tests/models/test_model_forward_no_implicit_sync.py` — sync gate,
  **8/8 PASSED**. The allowlist is AST-function-keyed, so the line shifts from
  regeneration don't matter; observed-sync counts are unchanged (the precompute
  fast path and the runtime fallback both behave as before). This PR also adds
  a `qwen3_5_vl-sdpa` case to the gate: the existing qwen3_5 cases are
  text-only sub-configs that never reach the vision tower, so the qwen3_5-family
  ViT precompute consumer was previously ungated.
- `pytest tests/models/test_models_logits_equal_v5.py -k qwen` — **16/16
  PASSED**, no numeric regression across the qwen2.5 / qwen3 multimodal lineage.
- `pytest tests/data/test_mm_metadata.py` — **4/4 PASSED** (rewritten for the
  collator-hook protocol: hook handoff, `preforward` nested-dict device move,
  hook picklability + temporal-unroll correctness).
- `make quality` — passes.
- `python -m veomni.patchgen.check_patchgen` — all generated files in sync.

### API and Usage Example

No public API change. Internal:

- `ViT.forward` now takes a single optional `vit_metadata` dict kwarg instead
  of 3 `vit_*` kwargs. All optional, all with runtime fallback.
- `veomni/data/mm_metadata.py` removed; its `per_sample_metadata` /
  `merge_position_id_returns` helpers are gone (the first deleted, the second
  inlined).

The full producer/consumer contract is in `.agents/knowledge/multimodal_metadata.md`.

### Design & Code Changes

- `veomni/data/mm_metadata.py` — **deleted**.
- `veomni/data/data_transform.py` — transforms emit `*_grid_thw` tensors +
  `position_ids` (inlined); no `per_sample_metadata` / `merge_position_id_returns`.
- `veomni/data/data_collator.py` — `PackingCollator` drops the `*_grid_thw_list`
  flatten special-case.
- `veomni/models/transformers/{qwen3_vl,qwen3_vl_moe,qwen3_5,qwen3_5_moe,qwen3_omni_moe}/*_gpu_patch_gen_config.py`
  — `collate_multimodal_metadata` reads the packed grid tensor; `Model.forward`
  builds the `vit_metadata` sub-dict; `ViT.forward` / `dummy_forward` consume it.
- `veomni/models/transformers/{qwen2_5vl,qwen2_5_omni}/*_gpu_patch_gen_config.py`
  — drain the renamed `vit_metadata` kwarg (fallback-only).
- All affected `generated/` files regenerated via patchgen.
- `tests/data/test_mm_metadata.py` — rewritten for the collator-hook protocol.
- `tests/models/test_model_forward_no_implicit_sync.py` — added the
  `qwen3_5_vl-sdpa` case so the qwen3_5-family ViT is gated for syncs
  (SDPA because FA2 NaNs on the qwen3_5 toy config; ViT metadata syncs are
  attention-implementation-independent). Two allowlist entries: the non-FA
  `Qwen3_5VisionAttention.forward` varlen split (`HF-eager-only` — production
  uses FA2) and `get_rope_index` (algorithm-essential mrope).
- `.agents/knowledge/multimodal_metadata.md` — rewritten for the single-key
  contract; added an "Adding the hook to a new model" checklist.
- `.agents/skills/veomni-new-model/SKILL.md` + `veomni-migrate-transformers-v5/SKILL.md`
  — added the multimodal-metadata-hook checklist.

### Follow-up work (not in this PR)

- Wire `qwen2_vl` (no metadata hook yet).
- Full `qwen2_5_vl` / `qwen2_5_omni` ViT integration — their window-attention
  layout derives `cu_window_seqlens` from `get_window_index` (config-dependent),
  so the hook needs a `partial(..., window_size=...)`; needs GPU
  `logits_equal_v5` verification.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks
- [x] Added/updated documentation (`.agents/knowledge/multimodal_metadata.md` + 2 skills)
- [x] Added tests to CI workflow (`tests/data/test_mm_metadata.py` rewritten; sync gate + logits_equal_v5 already wired into `gpu_unit_tests_v5`)

### CODE REVIEW

File: veomni/models/transformers/qwen3_vl_moe/generated/patched_modeling_qwen3_vl_moe_gpu.py [hunk 1]
```diff
@@ -268,21 +268,22 @@ def collate_multimodal_metadata(batch, sp_pad):
...
+    # Temporal unroll: each (t, h, w) expands to ``t`` cu steps of ``h * w``.
+    for modality, grid_key, pad_key in (
+        ("image", "image_grid_thw", "pixel_values"),
+        ("video", "video_grid_thw", "pixel_values_videos"),
     ):
-        grid_list = md.get(list_key)
+        grid = batch.get(grid_key)
+        if grid is None:
+            continue
+        grid_list = grid.tolist() if torch.is_tensor(grid) else grid
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_vl_moe/generated/patched_modeling_qwen3_vl_moe_gpu.py:285]: same

File: veomni/models/transformers/qwen3_vl/generated/patched_modeling_qwen3_vl_gpu.py [hunk 2]
```diff
@@ -237,21 +237,22 @@ def collate_multimodal_metadata(batch, sp_pad):
...
+    # Temporal unroll: each (t, h, w) expands to ``t`` cu steps of ``h * w``.
+    for modality, grid_key, pad_key in (
+        ("image", "image_grid_thw", "pixel_values"),
+        ("video", "video_grid_thw", "pixel_values_videos"),
     ):
-        grid_list = md.get(list_key)
+        grid = batch.get(grid_key)
+        if grid is None:
+            continue
+        grid_list = grid.tolist() if torch.is_tensor(grid) else grid
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_vl/generated/patched_modeling_qwen3_vl_gpu.py:254]: same

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py [hunk 3]
```diff
@@ -154,21 +154,22 @@ def collate_multimodal_metadata(batch, sp_pad):
...
+    # Temporal unroll: each (t, h, w) expands to ``t`` cu steps of ``h * w``.
+    for modality, grid_key, pad_key in (
+        ("image", "image_grid_thw", "pixel_values"),
+        ("video", "video_grid_thw", "pixel_values_videos"),
     ):
-        grid_list = md.get(list_key)
+        grid = batch.get(grid_key)
+        if grid is None:
+            continue
+        grid_list = grid.tolist() if torch.is_tensor(grid) else grid
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py:169]: The collate_multimodal_metadata hook is a metadata_collate_func — it runs inside the collator, which executes in CPU dataloader workers strictly before trainer.preforward moves the batch to the device.

**Review by Luosuu — APPROVED** [review]
